src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.Windows.cs

```diff
@@ -299,8 +308,80 @@ public ApartmentState GetApartmentState()
             }
         }
 
-        // TODO: https://github.com/dotnet/corefx/issues/20766
-        public bool TrySetApartmentState(ApartmentState state) { throw new PlatformNotSupportedException(); }
+        public bool TrySetApartmentState(ApartmentState state)
+        {
+            if (this != CurrentThread)
+            {
+                using (LockHolder.Hold(_lock))
+                {
+                    if (HasStarted())
+                        throw new ThreadStateException();
+                    _initialAppartmentState = state;
+                    return true;
+                }
+            }
+
+            if (state != ApartmentState.Unknown)
+            {
+                InitializeCom(state);
+            }
+            else
+            {
+                UninitializeCom();
+            }
+
+            // Clear the cache and check whether new state matches the desired state
+            t_apartmentType = ApartmentType.Unknown;
+            return state == GetApartmentState();
+        }
+
+        private void InitializeComOnNewThread()
+        {
+            InitializeCom(_initialAppartmentState);
+        }
+
+        internal static void InitializeCom(ApartmentState state = ApartmentState.MTA)
+        {
+            if (t_comInitialized)
+                return;
+
+#if ENABLE_WINRT
+            int hr = Interop.WinRT.RoInitialize(
+                (state == ApartmentState.STA) ? Interop.WinRT.RO_INIT_SINGLETHREADED
+                    : Interop.WinRT.RO_INIT_MULTITHREADED);
+#else
+            int hr = Interop.Ole32.CoInitializeEx(IntPtr.Zero,
+                (state == ApartmentState.STA) ? Interop.Ole32.COINIT_APARTMENTTHREADED
+                    : Interop.Ole32.COINIT_MULTITHREADED);
+#endif
+            // RPC_E_CHANGED_MODE indicates this thread has been already initialized with a different
+            // concurrency model.
+            if (hr == HResults.RPC_E_CHANGED_MODE)
+                return;
```

Comment:
I assume that returning from this function here with t_comInitialized set to false is intentional, right?
One potential side effect of this is that calling TrySetApartmentState(ApartmentType.Unknown) will have no effect and will not call Unintialize(). Though I can see that it may actually be the desired behavior in some cases.

Reply:
Thanks!